flink-runtime/src/test/java/org/apache/flink/runtime/testutils/InMemoryReporter.java

```diff
@@ -159,17 +168,51 @@ void applyRemovals() {
 
     @Override
     public void notifyOfAddedMetric(Metric metric, String metricName, MetricGroup group) {
-        MetricGroup metricGroup = unwrap(group);
-        LOG.debug("Registered {} @ {}", metricName, metricGroup);
-        synchronized (this) {
-            metrics.computeIfAbsent(metricGroup, dummy -> new HashMap<>()).put(metricName, metric);
+        if (captureMode != CaptureMode.NONE) {
+            MetricGroup metricGroup = unwrap(group);
+            LOG.debug("Registered {} @ {}", metricName, metricGroup);
+            synchronized (this) {
+                metrics.computeIfAbsent(metricGroup, dummy -> new HashMap<>())
+                        .put(metricName, metric);
+            }
         }
     }
 
     @Override
     public void notifyOfRemovedMetric(Metric metric, String metricName, MetricGroup group) {
-        synchronized (this) {
-            removedGroups.add(unwrap(group));
+        if (captureMode != CaptureMode.NONE) {
```

Comment:
I'm not particularly happy with all the `if-else` but I haven't found a way to express it without severely overengineering. @tillrohrmann any ideas?